Document anomalies under read-committed. (#1601)
neo4j/docs-operations: modules/ROOT/pages/database-internals/concurrent-data-access.adoc

=== Non-repeatable reads

A non-repeatable read is when the same transaction reads the same data but gets inconsistent results.
This can easily happen if reading the same data twice in a query and the data gets modified in-between by another concurrent query.

.Non-repeatable read
====
The following example query shows that reading the same property twice can give inconsistent results.
If there are other queries running concurrently, it is not guaranteed that `p1` and `p2` have the same value.

[source, cypher, role="noheader"]
----
MATCH (n:Example {id: 42})
WITH n.prop AS p1
// another concurrent query changes the value of n.prop here.
WITH *, n.prop AS p2
RETURN p1, p2
----
====

The easiest way to work around this is to only read each property once, and keep it as long as needed in the query.

=== Missing and double reads

When scanning an xref:performance/index-configuration.adoc[index], entities may be observed multiple times or skipped entirely, even if they are present in the index.
This is true even for indexes that back link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/constraints/examples/#constraints-examples-node-uniqueness[uniqueness constraints].

> During the scan, if another concurrent query changes a property of an entity in the index, the entity might appear again if the property was changed to a position ahead of the scan.

Reviewer: ```suggestion
During the scan, if another concurrent query changes an entity's property to a position ahead of the scan, the entity might appear again in the index.
```

> Similarly, the entity might not appear at all if the property changed to a position that has already been scanned.

Reviewer: ```suggestion
Similarly, the entity may not appear at all if the property is changed to a previously scanned position.
```

This anomaly can only occur with operators that scan an index, or parts of an index, for example link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/planning-and-tuning/operators/operators-detail/#query-plan-node-index-scan[`NodeIndexScan`] or link:{neo4j-docs-base-uri}/cypher-manual/{page-version}/planning-and-tuning/operators/operators-detail/#query-plan-directed-relationship-index-seek-by-range[`DirectedRelationshipIndexSeekByRange`].

.Missing and double read
====

> In the following query one could expect to see every node `n` that has the property `prop` exactly once.

Reviewer: ```suggestion
In the following query you can expect to see every node `n` that has the property `prop` exactly once.
```

Author: I wanted to point out that "seeing every node `n` that has the property `prop` exactly once" could be considered a reasonable assumption. But, as it turns out, that assumption does not hold.

Using "you can expect" delivers the wrong message, doesn't it. I was also not entirely happy with "one could expect". Do you maybe have another suggestion?

> However, concurrent updates that change the `prop` property while the index is being scanned, can lead to a node appearing multiple times or not at all in the result set.

Reviewer: ```suggestion
However, concurrent updates that modify the `prop` property during index scanning may cause a node to appear multiple times or not at all in the result set.
```
